(window.webpackJsonp=window.webpackJsonp||[]).push([[10],{140:function(e,n,t){"use strict";t.r(n),t.d(n,"frontMatter",(function(){return l})),t.d(n,"rightToc",(function(){return o})),t.d(n,"default",(function(){return u}));t(51),t(24),t(19),t(20),t(52),t(0);var r=t(166);function a(){return(a=Object.assign||function(e){for(var n=1;n<arguments.length;n++){var t=arguments[n];for(var r in t)Object.prototype.hasOwnProperty.call(t,r)&&(e[r]=t[r])}return e}).apply(this,arguments)}var l={id:"tryto",title:"TryTo",sidebar_label:"TryTo"},o=[{value:"Construction",id:"construction",children:[]},{value:"Failure",id:"failure",children:[]},{value:"Recovery",id:"recovery",children:[]},{value:"Methods",id:"methods",children:[]},{value:"Examples",id:"examples",children:[]}],i={rightToc:o},c="wrapper";function u(e){var n=e.components,t=function(e,n){if(null==e)return{};var t,r,a={},l=Object.keys(e);for(r=0;r<l.length;r++)t=l[r],n.indexOf(t)>=0||(a[t]=e[t]);return a}(e,["components"]);return Object(r.b)(c,a({},i,t,{components:n,mdxType:"MDXLayout"}),Object(r.b)("p",null,"The ",Object(r.b)("inlineCode",{parentName:"p"},"TryTo")," control gives us the ability to write safe code without focusing on try-catch blocks in the presence of exceptions.\n",Object(r.b)("inlineCode",{parentName:"p"},"TryTo")," wraps given ",Object(r.b)("inlineCode",{parentName:"p"},"callable")," and catch any ",Object(r.b)("inlineCode",{parentName:"p"},"Throwable")," that will occur while executing it."),Object(r.b)("p",null,"Remember that ",Object(r.b)("inlineCode",{parentName:"p"},"TryTo")," will may re-throw exception when calling ",Object(r.b)("inlineCode",{parentName:"p"},"get()")," method. You can prevent from this using ",Object(r.b)("inlineCode",{parentName:"p"},"isSuccess()")," or ",Object(r.b)("inlineCode",{parentName:"p"},"isFailure()")," methods. "),Object(r.b)("h2",{id:"construction"},"Construction"),Object(r.b)("p",null,"Use ",Object(r.b)("inlineCode",{parentName:"p"},"run")," static method:"),Object(r.b)("pre",null,Object(r.b)("code",a({parentName:"pre"},{className:"language-php"}),"/** @var TryTo<Result> $result */\n$result = TryTo::run(function(){throw new \\DomainException('use ddd');});\n\n$result->getOrElse(new Result());\n")),Object(r.b)("h2",{id:"failure"},"Failure"),Object(r.b)("p",null,"In case of an exception, ",Object(r.b)("inlineCode",{parentName:"p"},"TryTo")," returns an instance of ",Object(r.b)("inlineCode",{parentName:"p"},"Failure"),".\nYou can use ",Object(r.b)("inlineCode",{parentName:"p"},"getCause(): Throwable")," method to gets the underlying exception."),Object(r.b)("pre",null,Object(r.b)("code",a({parentName:"pre"},{className:"language-php"}),"$result = TryTo::run(function(){throw new \\DomainException('use ddd');});\n\n$result->getCause(); // returns instance of DomainException\n")),Object(r.b)("h2",{id:"recovery"},"Recovery"),Object(r.b)("p",null,"Common use case is recovery from failure. ",Object(r.b)("inlineCode",{parentName:"p"},"TryTo")," gives you interesting option for that ",Object(r.b)("inlineCode",{parentName:"p"},"recover(string $throwable, callable $recovery): self")),Object(r.b)("pre",null,Object(r.b)("code",a({parentName:"pre"},{className:"language-php"}),"$value = TryTo::run(function () {throw new \\DomainException('use ddd'); })\n        ->recover(\\RuntimeException::class, function () {return 'runtime handled'; })\n        ->recover(\\DomainException::class, function () {return 'domain handled'; })\n        ->get();\n// $value === 'domain handled'\n")),Object(r.b)("h2",{id:"methods"},"Methods"),Object(r.b)("ul",null,Object(r.b)("li",{parentName:"ul"},Object(r.b)("inlineCode",{parentName:"li"},"andThen(callable $callable): self")," - allow to chain next operation"),Object(r.b)("li",{parentName:"ul"},Object(r.b)("inlineCode",{parentName:"li"},"andFinally(callable $callable): self")," - run given $callable whatever the result is"),Object(r.b)("li",{parentName:"ul"},Object(r.b)("inlineCode",{parentName:"li"},"onSuccess(callable $consumer): self")," - run given $consumer with success value as an argument"),Object(r.b)("li",{parentName:"ul"},Object(r.b)("inlineCode",{parentName:"li"},"onFailure(callable $consumer): self")),Object(r.b)("li",{parentName:"ul"},Object(r.b)("inlineCode",{parentName:"li"},"onSpecificFailure(string $throwable, callable $consumer): self"))),Object(r.b)("h2",{id:"examples"},"Examples"),Object(r.b)("p",null,"Try to run operation, log error, and get default value - all in one simple object-oriented approach:"),Object(r.b)("pre",null,Object(r.b)("code",a({parentName:"pre"},{className:"language-php"}),"$result = TryTo::run(function () {\n    return 'success';\n})->onFailure(function (\\Throwable $throwable) {\n    log($throwable);\n})->getOrElse('default');\n")))}u.isMDXComponent=!0},166:function(e,n,t){"use strict";t.d(n,"a",(function(){return i})),t.d(n,"b",(function(){return s}));var r=t(0),a=t.n(r),l=a.a.createContext({}),o=function(e){var n=a.a.useContext(l),t=n;return e&&(t="function"==typeof e?e(n):Object.assign({},n,e)),t},i=function(e){var n=o(e.components);return a.a.createElement(l.Provider,{value:n},e.children)};var c="mdxType",u={inlineCode:"code",wrapper:function(e){var n=e.children;return a.a.createElement(a.a.Fragment,{},n)}},p=Object(r.forwardRef)((function(e,n){var t=e.components,r=e.mdxType,l=e.originalType,i=e.parentName,c=function(e,n){var t={};for(var r in e)Object.prototype.hasOwnProperty.call(e,r)&&-1===n.indexOf(r)&&(t[r]=e[r]);return t}(e,["components","mdxType","originalType","parentName"]),p=o(t),s=r,b=p[i+"."+s]||p[s]||u[s]||l;return t?a.a.createElement(b,Object.assign({},{ref:n},c,{components:t})):a.a.createElement(b,Object.assign({},{ref:n},c))}));function s(e,n){var t=arguments,r=n&&n.mdxType;if("string"==typeof e||r){var l=t.length,o=new Array(l);o[0]=p;var i={};for(var u in n)hasOwnProperty.call(n,u)&&(i[u]=n[u]);i.originalType=e,i[c]="string"==typeof e?e:r,o[1]=i;for(var s=2;s<l;s++)o[s]=t[s];return a.a.createElement.apply(null,o)}return a.a.createElement.apply(null,t)}p.displayName="MDXCreateElement"}}]);